backend_kvm.go

```diff
@@ -74,14 +75,58 @@ func (b kvmBackend) hostKernelRelease() (string, error) {
 	return "", fmt.Errorf("No kernel found")
 }
 
+func (b kvmBackend) hostKernelPath(kernelRelease string) (string, error) {
+	kernelDir := "/boot"
+	kernelPrefix := "vmlinuz-"
+
+	/* First, try to find a kernel with a well-known name */
+	kernelPath := filepath.Join(kernelDir, kernelPrefix + kernelRelease)
+	if _, err := os.Stat(kernelPath); err == nil {
+		return kernelPath, nil
+	}
+
+	/* Otherwise, inspect each kernel installed, and look for the release
+	 * string straight in the binary. Not pretty, but it works. */
+	needle := kernelRelease
+	if !strings.HasSuffix(needle, " ") {
+		needle += " "
+	}
+
+	files, err := ioutil.ReadDir(kernelDir)
+	if err != nil {
+		return "", err
+	}
+
+	for _, f := range files {
+		if !strings.HasPrefix(f.Name(), kernelPrefix) || f.IsDir() {
+			continue
+		}
+
+		kernelPath := filepath.Join(kernelDir, f.Name())
+		buf, err := ioutil.ReadFile(kernelPath)
+		if err != nil {
+			fmt.Fprintln(os.Stderr, "Failed to read kernel:", err)
+			continue
+		}
+
+		if !bytes.Contains(buf, []byte(needle)) {
```

Comment:
that's.. interesting. Is there no existing golang library that can disect a vmlinuz file? e.g. file can:
```
$  file /boot/vmlinuz-5.10.0-3-amd64 
/boot/vmlinuz-5.10.0-3-amd64: Linux kernel x86 boot executable bzImage, version 5.10.0-3-amd64 (debian-kernel@lists.debian.org) #1 SMP Debian 5.10.17-1 (2021-02-12), RO-rootFS, swap_dev 0x6, Normal VGA
```

Reply:
I couldn't find any golang library for this, IIRC neither did Arnaud (original author). The Arch initrd generator `mkinitcpio` does _this exact_ search within vmlinuz and from a quick look so does `file` itself. Generally I would be weary in parsing the output of `file`. Especially since the `file` `cpio` (variant) detection has been broken for years.

From my Arch box:
```
/boot/vmlinuz-linux: Linux kernel x86 boot executable bzImage, version 5.12.6-arch1-1 (linux@archlinux) #1 SMP PREEMPT Sun, 23 May 2021 00:45:50 +0000, RO-rootFS, swap_dev 0x8, Normal VGA
/boot/vmlinuz-linux-lts: Linux kernel x86 boot executable bzImage, version 5.10.39-1-lts (linux-lts@archlinux) #1 SMP Sat, 22 May 2021 10:57:31 +0000, RO-rootFS, swap_dev 0x8, Normal VGA
```

Reply:
file know how to parse x86 kernels; See https://www.kernel.org/doc/Documentation/x86/boot.txt for the header it parses (real mode header). 

As this only happens on machine without predictable names i'm not entirely against it but ugh..